tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/access/FieldAccessTest.java

```diff
@@ -79,38 +79,56 @@ public void testNegativeCases() {
-        validateError(negativeResult, i++, "field access can only be used to access required fields, field 'c' is " +
-                "optional in record(s) 'TA'", 249, 17);
+        validateError(negativeResult, i++, "field access can only be used to access required fields or optional " +
+                "fields of non-nilable types, field 'x' is undeclared in record(s) 'SA', 'UA', and 'VA'", 251, 17);
+        validateError(negativeResult, i++, "field access can only be used to access required fields or optional " +
+                "fields of non-nilable types, field 'y' is undeclared in record(s) 'SA', and 'UA'", 252, 17);
+        validateError(negativeResult, i++, "field access can only be used to access required fields or optional " +
+                "fields of non-nilable types, field 'z' is undeclared in record(s) 'TA'", 253, 17);
 
-        validateError(negativeResult, i++, "field access can only be used to access required fields, field 'x' is " +
-                "undeclared in record(s) 'SA', 'UA', and 'VA'", 251, 17);
-        validateError(negativeResult, i++, "field access can only be used to access required fields, field 'y' is " +
-                "undeclared in record(s) 'SA', and 'UA'", 252, 17);
-        validateError(negativeResult, i++, "field access can only be used to access required fields, field 'z' is " +
-                "undeclared in record(s) 'TA'", 253, 17);
+        validateError(negativeResult, i++, "field access can only be used to access required fields or optional " +
+                "fields of non-nilable types, field 'x' is undeclared in record(s) 'RB', 'TB', and 'UB'",
+                293, 17);
+        validateError(negativeResult, i++, "field access can only be used to access required fields or optional " +
+                "fields of non-nilable types, field 'y' is undeclared in record(s) 'QB', and 'TB'",
+                294, 17);
+        validateError(negativeResult, i++, "field access can only be used to access required fields or optional " +
+                "fields of non-nilable types, field 'z' is undeclared in record(s) 'RB', 'SB', and 'VB'",
+                295, 17);
 
-        validateError(negativeResult, i++, "field access can only be used to access required fields, field 'x' is " +
-                "undeclared in record(s) 'RB', 'TB', and 'UB' and optional in record(s) 'QB', 'SB', and 'VB'", 293, 17);
-        validateError(negativeResult, i++, "field access can only be used to access required fields, field 'y' is " +
-                "undeclared in record(s) 'QB', and 'TB' and optional in record(s) 'RB', 'UB', and 'VB'", 294, 17);
-        validateError(negativeResult, i++, "field access can only be used to access required fields, field 'z' is " +
-                "undeclared in record(s) 'RB', 'SB', and 'VB' and optional in record(s) 'QB', and 'TB'", 295, 17);
+        validateError(negativeResult, i++, "field access can only be used to access required fields or optional " +
```

Comment:
We need a separate error for unions, right?

1. for non-union cases, requirement is either that the field is required (type may be nilable) or that the field is optional and is of a non-nil type
2. for union cases, requirement is that in each member the field is either required or optional, and the type is not nilable

Here `field access can only be used to access required fields or optional fields of non-nilable types, type of field 'i' includes nil in record(s) 'BarOne'` may be misleading, since `i` is a required field in `BarOne`.

```ballerina
type BarOne record {|
    int? i;
|};
```